### ISSUE TITLE
Codacy

### ISSUE THREAD

**Issue by AlexandruValeanu** [jgrapht/jgrapht#688]
I've seen [codacy](https://www.codacy.com/) (code coverage with a nice interface) used in a large number of open source projects. Do you think it's something that might be interesting for JGraphT?

**Comment by jsichi** [comment]
Yes, that would be awesome.

**Comment by AlexandruValeanu** [comment]
I can look into it and probably try in my own fork.

**Comment by jsichi** [comment]
Cool.  Once you get it working, I can help with setting up the access tokens for the master repository.

**Comment by AlexandruValeanu** [comment]
I got it working but I removed jgrapht from codacy since I do not have admin rights and they are requested in order to set the post-commit hook.

Steps for an admin to follow:

1. Log in into https://www.codacy.com/
2. Add `jgrapht` using `Add project`
3. Go to `My Projects -> jgrapht -> Settings -> Integrations`
4. Tick PR status/comment/summary. Check [this](https://support.codacy.com/hc/en-us/articles/207280219-GitHub-Integration) to see what they are.
5. Tick Hook status to set a pos-commit hook.

Codacy can mark a commit/PR as `fail` if it doesn't meet certain quility criteria. For now I suggest we use the standard codacy criteria ([more info about quality settings](https://support.codacy.com/hc/en-us/articles/360009164573-Quality-Settings)).
 
Note: You can see an example PR code review performed on #687. 

I suggest we start with all checks/reviews performed and later remove them if it gets too verbose.

Any suggestions/opinions are welcomed.

**Comment by jkinable** [comment]
I've getting quite a bit of noice from Codacy. Especially on larger PRs
codacy seems to send a flood of warnings with several (undesired)
suggestions. Should we start of with a more restricted set of rules for
Codacy?

**Comment by jsichi** [comment]
Right, let's leave it disabled for now until we've agreed on a non-noisy rule set.  

@AlexandruValeanu when you mentioned coverage, what I was actually hoping for was a test coverage result (run through all tests and detect functional code which isn't exercised).  I think that has higher value than static code analysis.  Is it possible to get that working?

**Comment by AlexandruValeanu** [comment]
It is possible using [codecov](https://codecov.io). You can see how a PR will look like [here](https://github.com/yoda-pa/yoda/pull/185).

**Comment by jsichi** [comment]
Yeah that looks good.  @d-michail @jkinable do you think these test coverage reports would be useful?  For new contributions, they should allow us to see at a glance how thoroughly the contributor approached testing.

**Comment by d-michail** [comment]
We can give it a try and see whether it is too much spam or not. 

The codacy one reminded me of sonarqube, where after a while the code gets filled with useless annotations so that sonarqube does not complain. :)

**Comment by jkinable** [comment]
We can certainly give it a try. It's always nice if a tool is able to lower our review burden.

Btw, is there a minimal set of rules that can be enforced by Codacy, i.e. a set of issues that should certainly be resolved by the user? I did like its integration into the github reviewing process and how it reported issues. The only problem was that it complained a little too much.
